lib/xwax/pitch_kalman.h

```diff
@@ -0,0 +1,325 @@
+
+    struct kalman_coeffs stable;
+    struct kalman_coeffs medium;
+    struct kalman_coeffs reactive;
+};
+
+/*
+ * Retune noise sensitivity without resetting state
+ */
+
+static inline void kalman_tune_sensitivity(struct pitch_kalman* p, struct kalman_coeffs* coeffs) {
+    if (!p || !coeffs) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    p->coeffs = coeffs;
+}
+
+/*
+ * Initialize the filter with dt, the coefficients and thresholds for
+ * the different mode switches
+ *
+ * q: acceleration noise spectral density (tune up if motion is more jittery)
+ * r: variance of dx measurement (tune up if observations are noisier)
+ */
+
+static inline void pitch_kalman_init(struct pitch_kalman* p, double dt,
+        struct kalman_coeffs stable, struct kalman_coeffs medium, struct kalman_coeffs reactive,
+        double medium_threshold, double scratch_threshold)
+{
+    if (!p || medium_threshold > scratch_threshold) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    /* Indexes for the matrices */
+
+    const size_t x = 0;
+    const size_t v = 1;
+
+    /* Sampling interval */
+
+    p->dt = dt;
+
+    /* State */
+
+    p->x = 0.0;
+    p->v = 0.0;
+
+    /* Start with large uncertainty so the filter trusts early measurements */
+
+    p->P[x][x] = 1e6;
+    p->P[x][v] = 0.0;
+    p->P[v][v] = 1e6;
+
+    /* Fixed thresholds for the mode switches */
+
+    p->scratch_threshold = scratch_threshold;
+    p->medium_threshold = medium_threshold;
+
+    /* Q and R for the different modes */
+
+    p->stable = stable;
+    p->medium = medium;
+    p->reactive = reactive;
+
+    /* Initialize as reactive */
+
+    kalman_tune_sensitivity(p, &p->reactive);
+}
+
+/*
+ * Feed one observation: in the last dt seconds, position moved by dx
+ */
+
+static inline void pitch_kalman_update(struct pitch_kalman* p, double dx) {
+
+    if (!p) {
+        errno = EINVAL;
+        perror(__func__);
+        return;
+    }
+
+    /* Sampling interval */
+
+    const double dt = p->dt;
+
+    /* Indexes for the matrices */
+
+    const size_t x = 0;
+    const size_t v = 1;
+
+    /*
+     * =================== Prediction step =====================
+     *
+     * State transition for constant-velocity model:
+     *   F = [ 1  dt
+     *         0  1 ]
+     *
+     * Discretizing, the process noise covariance is:
+     *   Q = Q * [ dt^3/3  dt^2/2
+     *             dt^2/2  dt     ]
+     *
+     * Here p->coeffs->Q is 'Q', the acceleration noise variance.
+     * Q[v][x] is not needed for the computations, therefore we put 0.0.
+     */
+
+    const double F[2][2] = { { 1.0,   dt },
+                             { 0.0,  1.0 }};
+
+    const double Q[2][2] = { { p->coeffs->Q * (pow(dt, 3) / 3.0),  p->coeffs->Q * (pow(dt, 2) / 2.0) },
+                             {                               0.0,  p->coeffs->Q * dt                 }};
```

Comment:
the previous version with `dt2` and `dt3` was IMO easier to read and faster (avoid calling into `cmath` in hot code if its reasonable to avoid).

Reply:
It's a draft